File: src/__tests__/mantle/testActionAndState.ts
import { ApplyActionFunc, ActionCommon, ReverseActionFunc } from '../../mantle/UndoRedoHistory';

export type TestState = {
  valueNumber: number;
  valueString: string;
};

export type TestAction = ActionCommon & (
  {
    type: 'INCREMENT_NUMBER' | 'DECREMENT_NUMBER';
    difference: number;
  } | {
    type: 'SET_STRING';
    oldValue: string;
    newValue: string;
  }
);

export const applyAction: ApplyActionFunc<TestAction, TestState> = action => state => {
  switch (action.type) {
    case 'INCREMENT_NUMBER': return { ...state, valueNumber: state.valueNumber + action.difference }
    case 'DECREMENT_NUMBER': return { ...state, valueNumber: state.valueNumber - action.difference }
    case 'SET_STRING': return { ...state, valueString: action.newValue }
    default: return state;
  }
}

export const reverseAction: ReverseActionFunc<TestAction> = action => {
  switch (action.type) {
    case 'INCREMENT_NUMBER': return { ...action, type: 'DECREMENT_NUMBER' };
    case 'DECREMENT_NUMBER': return { ...action, type: 'INCREMENT_NUMBER' };
    case 'SET_STRING': return { ...action, oldValue: action.newValue, newValue: action.oldValue };
    default: return action;
  }
};

File: npbloom-core/src/commonTest/kotlin/space/yuvalinguist/npbloom/history/TestActionAndState.kt
package space.yuvalinguist.npbloom.history

data class TestState(
    val numberValue: Int,
    val stringValue: String,
)

sealed interface TestAction : UndoableActionBase

data class IncrementNumber(val difference: Int) : TestAction
data class DecrementNumber(val difference: Int) : TestAction
data class SetString(val oldValue: String, val newValue: String) : TestAction

fun testApplyAction(state: TestState, action: TestAction) =
    when (action) {
        is DecrementNumber -> state.copy(numberValue = state.numberValue - action.difference)
        is IncrementNumber -> state.copy(numberValue = state.numberValue + action.difference)
        is SetString -> state.copy(stringValue = action.newValue)
    }

fun testReverseAction(action: TestAction) =
    when (action) {
        is DecrementNumber -> IncrementNumber(action.difference)
        is IncrementNumber -> DecrementNumber(action.difference)
        is SetString -> action.copy(oldValue = action.newValue, newValue = action.oldValue)
    }
